Objects/moduleobject.c

```diff
@@ -841,22 +841,26 @@ _Py_module_getattro_impl(PyModuleObject *m, PyObject *name, int suppress)
         }
         if (suppress != 1) {
             if (_PyModuleSpec_IsInitializing(spec)) {
-                origin = PyObject_GetAttr(spec, &_Py_ID(origin));
-                if (origin == NULL || origin == Py_None) {
-                    PyErr_Format(PyExc_AttributeError,
-                                "partially initialized "
-                                "module '%U' has no attribute '%U' "
-                                "(most likely due to a circular import)",
-                                mod_name, name);
+                int valid_spec = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
```

Comment:
When `PyObject_GetOptionalAttr()` returns -1, you should simply return NULL. It may be some unexpected error, from KeyboardIterrupt to MemoryError or RecursionError, and it should not be overridden by AttributeError.

The fact that it is overridden by AttributeError in the current code is a bug.

Reply:
It is a complex issue, see #112660.